### PR TITLE
[NPU]grok2 model support

### PR DESCRIPTION
**Pull request by KnightLTC**
## Motivation

grok2 model npu adapt

## Modifications

1. Grok2 employs softcap attention, but the existing attention operators in Ascend do not support softcap. To enable its functionality, we extended the native code of Torch's scaled_dot_product_attention to include softcap support
3. On NPU, the current rope calls the mrope operator, which has accuracy loss under BF16/FP16, resulting in a test dataset accuracy of only 0.2. Therefore, the input cast needs to be FP32
4. update forward triton code in grok for npu

## Accuracy Tests

image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
5. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
6. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

**Comment by ping1jing2** [comment]
Hi, @KnightLTC there are some suggestions:
1. please run `pre-commit` command before you push, see [contribution_guide.md](https://github.com/sgl-project/sglang/blob/main/docs/developer_guide/contribution_guide.md) for more details.
2. please add more description, at least including accuracy and preformance command and result.

File: python/sglang/srt/layers/rotary_embedding.py [hunk 1]
```diff
@@ -296,6 +296,10 @@ def forward_npu(
         ), "fused_set_kv_buffer_arg is not supported for npu implementation"
         if query.dtype == torch.bfloat16 and self.cos_sin_cache.dtype == torch.float:
             return self.forward_native(positions, query, key, offsets)
+        if get_bool_env_var("SGLANG_ENABLE_TORCH_COMPILE"):
```
Comment (xiaobaicxy) [inline python/sglang/srt/layers/rotary_embedding.py]: Using SGLANG_ENABLE_TORCH_COMPILE here is inappropriate

File: python/sglang/srt/layers/attention/torch_native_backend.py [hunk 2]
```diff
@@ -24,6 +25,51 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
...
+
+        attn_weight = query @ key.transpose(-2, -1) * scale_factor
+
+        if logit_cap > 0:
+            if logit_capping_method == "tanh":
+                attn_weight = logit_cap * torch.tanh(attn_weight / logit_cap)
+
+        attn_weight += attn_bias
+        attn_weight = torch.softmax(attn_weight, dim=-1)
+        attn_weight = torch.dropout(attn_weight, dropout_p, train=False)
```
Comment (xiaobaicxy) [inline python/sglang/srt/layers/attention/torch_native_backend.py]: There is usually no dropout during inference. Please delete this line of code

Comment (KnightLTC) [reply python/sglang/srt/layers/attention/torch_native_backend.py]: done

File: python/sglang/srt/layers/rotary_embedding.py (hunk 1 above)
Comment (KnightLTC) [reply python/sglang/srt/layers/rotary_embedding.py]: done

File: python/sglang/srt/models/grok.py [hunk 3]
```diff
@@ -162,8 +168,21 @@ def __init__(
...
+            router_logits = self.router_logit_softcapping * F.tanh(
+                router_logits / self.router_logit_softcapping
+            )
+            topk_output = self.topk(hidden_states, router_logits)
+
+            final_hidden_states = self.experts(hidden_states, topk_output)
+            return final_hidden_states.view(orig_shape)
+        else:
+            topk_output = self.topk(hidden_states, self.gate.weight)
+            return self.experts(hidden_states, topk_output)
```
Comment (iforgetmyname) [inline python/sglang/srt/models/grok.py:183]: this change makes non-sense

File: python/sglang/srt/models/grok.py [hunk 4]
```diff
@@ -337,7 +358,7 @@ def __init__(
         self.rope_theta = rope_theta
         rope_scaling = get_rope_scaling(config)
         self.load_presharded_attn = load_presharded_attn
-        self.alt_stream = alt_stream or torch.cuda.Stream()
+        self.alt_stream = alt_stream or torch.get_device_module().Stream()
```
Comment (iforgetmyname) [inline python/sglang/srt/models/grok.py]: revert

File: python/sglang/srt/models/grok.py [hunk 5]
```diff
@@ -451,7 +482,7 @@ def __init__(
         self.hidden_size = config.hidden_size
         self.residual_moe = getattr(config, "residual_moe", False)
         self.layer_id = layer_id
-        self.alt_stream = alt_stream or torch.cuda.Stream()
+        self.alt_stream = alt_stream or torch.get_device_module().Stream()
```
Comment (iforgetmyname) [inline python/sglang/srt/models/grok.py]: revert

File: python/sglang/srt/models/grok.py [hunk 6]
```diff
@@ -583,10 +614,10 @@ def forward(
 
     def moe_with_rmoe(self, x):
         if self.alt_stream is not None and get_is_capture_mode():
-            current_stream = torch.cuda.current_stream()
+            current_stream = torch.get_device_module().current_stream()
```
Comment (iforgetmyname) [inline python/sglang/srt/models/grok.py]: revert

File: python/sglang/srt/models/grok.py [hunk 7]
```diff
@@ -583,10 +614,10 @@ def forward(
...
             self.alt_stream.wait_stream(current_stream)
             mlp_result = self.mlp(x)
-            with torch.cuda.stream(self.alt_stream):
+            with torch.get_device_module().stream(self.alt_stream):
```
Comment (iforgetmyname) [inline python/sglang/srt/models/grok.py]: revert

File: python/sglang/srt/models/grok.py [hunk 8]
```diff
@@ -620,7 +651,7 @@ def __init__(
             prefix=add_prefix("embed_tokens", prefix),
         )
 
-        self.alt_stream = torch.cuda.Stream()
+        self.alt_stream = torch.get_device_module().Stream()
```
Comment (iforgetmyname) [inline python/sglang/srt/models/grok.py]: revert

File: test/registered/ascend/llm_models/test_ascend_grok_2.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+from sglang.test.ascend.gsm8k_ascend_mixin import GSM8KAscendMixin
+from sglang.test.ci.ci_register import register_npu_ci
+from sglang.test.test_utils import CustomTestCase
+
+register_npu_ci(est_time=400, suite="nightly-16-npu-a3", nightly=True)
+
+
+class TestGrok2(GSM8KAscendMixin, CustomTestCase):
+    model = "/root/.cache/modelscope/hub/models/huihui-ai/grok-2"
```
Comment (iforgetmyname) [inline test/registered/ascend/llm_models/test_ascend_grok_2.py:11]: this model is incorrect
